plumbing/object/commit_walker.go

```diff
@@ -181,3 +183,127 @@ func (w *commitPostIterator) ForEach(cb func(*Commit) error) error {
 }
 
 func (w *commitPostIterator) Close() {}
+
+// commitAllIterator stands for commit iterator for all refs.
+type commitAllIterator struct {
+	// el points to the current commit.
+	el *list.Element
+}
+
+// NewCommitAllIter returns a new commit iterator for all refs.
+// s is a repo Storer used to get commits and references.
+// fn is a commit iterator function, used to iterate through ref commits in chosen order
+func NewCommitAllIter(s storage.Storer, fn func(*Commit) CommitIter) (CommitIter, error) {
+	l := list.New()
+	m := make(map[plumbing.Hash]*list.Element)
+
+	// ...along with the HEAD
+	head, err := storer.ResolveReference(s, plumbing.HEAD)
+	if err != nil {
+		return nil, err
+	}
+	headCommit, err := GetCommit(s, head.Hash())
+	if err != nil {
+		return nil, err
+	}
+	err = fn(headCommit).ForEach(func(c *Commit) error {
+		el := l.PushBack(c)
+		m[c.Hash] = el
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	refIter, err := s.IterReferences()
+	if err != nil {
+		return nil, err
+	}
+	defer refIter.Close()
+	err = refIter.ForEach(func(r *plumbing.Reference) error {
+		if r.Hash() == head.Hash() {
+			// we already have the HEAD
+			return nil
+		}
+		c, _ := GetCommit(s, r.Hash())
+		// if it's not a commit - skip it.
+		if c == nil {
+			return nil
+		}
+
+		el, ok := m[c.Hash]
```

Comment:
This check may be done before `GetCommit` as `c.Hash` should be the same as `r.Hash()`.